jvm-packages/xgboost4j-spark-gpu/src/native/xgboost4j_spark.cpp

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstddef>
+#include <cstdint>
+#include <stdexcept>
+#include <stdlib.h>
+#include <string>
+#include <vector>
+#include <memory>
+#include <iostream>
+
+#include <cuda_runtime.h>
+
+#include "xgboost4j_spark_gpu.h"
+#include "xgboost4j_spark.h"
+
+#if defined(XGBOOST_USE_RMM) && XGBOOST_USE_RMM == 1
+#include "rmm/mr/device/per_device_resource.hpp"
+#include "rmm/mr/device/thrust_allocator_adaptor.hpp"
+#include "rmm/device_buffer.hpp"
+#endif  // defined(XGBOOST_USE_RMM) && XGBOOST_USE_RMM == 1
+
+struct gpu_column_data {
```

Comment:
I thought we have similar data structures in the xgboost4j package already?

Reply:
yeah, we need to convert jvm object to the c++ style.